datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -700,6 +701,288 @@ impl<'a, S: SimplifyInfo> ExprRewriter for Simplifier<'a, S> {
                 return Err(DataFusionError::ArrowError(ArrowError::DivideByZero));
             }
 
+            //
+            // Rules for BitwiseAnd
+            //
+
+            // A & null -> null
+            Expr::BinaryExpr(BinaryExpr {
+                left: _,
+                op: BitwiseAnd,
+                right,
+            }) if is_null(&right) => *right,
+
+            // null & A -> null
+            Expr::BinaryExpr(BinaryExpr {
+                left,
+                op: BitwiseAnd,
+                right: _,
+            }) if is_null(&left) => *left,
+
+            // A & 0 -> 0
+            Expr::BinaryExpr(BinaryExpr {
+                left: _,
+                op: BitwiseAnd,
+                right,
+            }) if is_zero(&right) => *right,
```

Comment:
I think this rule also has to check that `A` is non nullable as `null & 0` is `null` (not `0`, which is what this code does)

```sql
postgres=# select NULL & 0 ;
 ?column?
----------

(1 row)
```

The same comment applies to `0 & A` as well